runtime/compiler/runtime/JITServerAOTDeserializer.cpp

```diff
@@ -898,3 +901,600 @@ JITServerLocalSCCAOTDeserializer::updateSCCOffsets(SerializedAOTMethod *method,
+
+   auto it = _wellKnownClassesMap.find(record->id());
+   if (it != _wellKnownClassesMap.end())
+      {
+      if (revalidateWellKnownClasses(it->second, comp, wasReset))
+         {
+         return true;
+         }
+      else
+         {
+         TR::Compiler->persistentGlobalMemory()->freePersistentMemory(it->second);
+         it->second = NULL;
+         if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+               "Invalidated cached well-known classes record ID %zu", record->id()
+            );
+         return false;
+         }
+      }
+   isNew = true;
+
+   // Initialize the "well-known classes" object (see SymbolValidationManager::populateWellKnownClasses()).
+   TR_ASSERT(record->list().length() <= WELL_KNOWN_CLASS_COUNT, "Too many well-known classes");
+   // The first entry of the chain offsets contains the number of entries (and not the size in bytes as with class chains)
+   auto chainOffsets = (uintptr_t *)TR::Compiler->persistentGlobalMemory()->allocatePersistentMemory((1 + record->list().length()) * sizeof(uintptr_t));
+   chainOffsets[0] = record->list().length();
+   auto chainData = chainOffsets + 1;
+   // Get the class chain SCC offsets for each class chain ID (which should already be cached and (re)validated).
+   for (size_t i = 0; i < record->list().length(); ++i)
+      chainData[i] = encodeClassChainOffset(record->list().ids()[i]);
+
+   addToChainMap(_wellKnownClassesMap, it, record->id(), chainOffsets);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Cached well-known classes record ID %zu", record->id());
+   return true;
+   }
+
+bool
+JITServerNoSCCAOTDeserializer::cacheRecord(const ThunkSerializationRecord *record, TR::Compilation *comp, bool &isNew, bool &wasReset)
+   {
+   // Unlike the rest of the cacheRecord functions, we do not need to acquire a monitor here, as we can rely on
+   // the internal synchronization of getJ2IThunk and setJ2IThunk. Since we don't touch any internal caches, we can
+   // skip checking for a reset.
+
+   auto fej9vm = comp->fej9vm();
+   void *thunk = fej9vm->getJ2IThunk((char *)record->signature(), record->signatureSize(), comp);
+   if (thunk)
+      return true;
+   isNew = true;
+
+   TR::CompilationInfoPerThread *compInfoPT = fej9vm->_compInfoPT;
+   uint8_t *thunkStart = TR_JITServerRelocationRuntime::copyDataToCodeCache(record->thunkStart(), record->thunkSize(), fej9vm);
+   if (!thunkStart)
+      compInfoPT->getCompilation()->failCompilation<TR::CodeCacheError>("Failed to allocate space in the code cache");
+
+   auto thunkAddress = thunkStart + 8;
+   void *vmHelper = j9ThunkVMHelperFromSignature(fej9vm->_jitConfig, record->signatureSize(), (char *)record->signature());
+   compInfoPT->reloRuntime()->reloTarget()->performThunkRelocation(reinterpret_cast<uint8_t *>(thunkAddress), (uintptr_t)vmHelper);
```

Comment:
Will this thunk be relocated again during the AOT load? If so, is thunk relocation idempotent? Would be nice to add a comment explaining what exactly happens with the thunk after deserialization.